helix-term/src/commands/typed.rs

```diff
@@ -322,6 +322,52 @@ fn force_write(
     write_impl(cx, args.first(), true)
 }
 
+fn rename_impl(cx: &mut compositor::Context, new_name: Option<&Cow<str>>) -> anyhow::Result<()> {
+    let (_view, doc) = current!(cx.editor);
+
+    let path = if let Some(path) = doc.relative_path() {
+        path
+    } else {
+        return Err(anyhow!("File does not exist to rename."));
+    };
+
+    let mut new_path = path.clone();
+    new_path.set_file_name(OsStr::new(new_name.unwrap().as_ref()));
+
+    if new_path.parent() != path.parent() {
+        cx.editor.set_status("Root path must be same.");
+        return Ok(());
+    }
+
+    if new_path.exists() {
+        cx.editor.set_status("File already exist.");
+        return Ok(());
+    }
```

Comment:
These should be errors since they prevent the rename from happening:

```suggestion
    ensure!(new_path.parent() == path.parent(), "Root path must be the same.");

    ensure!(!new_path.exists(), "File already exists.");
```